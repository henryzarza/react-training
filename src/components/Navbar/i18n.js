import i18next from 'i18next';

i18next.addResources('en', 'NAVBAR', {
  DESCRIPTION:
    'A simple application using React (with hooks, portals, context, and between other interesting things), this app is just a demo becasue I wanna learn to connect GraphQL APIs to React apps.',
  REPOSITORY: 'Repository:',
  MADE_TEXT: 'Made with',
  MADE_BY: 'by Henry Zarza',
  HOME: 'Home',
  TIME_ZONES: 'Time zones',
  INTERESTING: 'Interesting',
});
