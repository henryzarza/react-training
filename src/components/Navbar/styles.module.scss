@import 'src/scss/variables';

.content {
  align-items: center;
  background: {
    image: url('../../assets/navbar.jpeg');
    position: center;
    repeat: no-repeat;
    size: cover;
  };
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: center;
  min-height: 40vh;
  padding: 1rem;
  position: relative;

  & > * {
    color: var(--navbar-text-color);
    z-index: 1;
  }

  &::before {
    background-color: rgba(0, 0, 0, 0.2);
    bottom: 0;
    content: '';
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
  }
}

.title {
  text-shadow: 0.1rem 0.1rem 0.2rem var(--shadow-btn);
}

.paragraph {
  max-width: 100%;
  width: 21.5rem;
}

.navbar {
  display: flex;
  position: sticky;
  top: 0;
  z-index: 1;
}

.navbar-item {
  background-color: var(--primary-color);
  color: var(--navbar-text-color);
  cursor: pointer;
  flex: 1;
  padding: 1.5rem 1rem;
  position: relative;
  transition: all $transition-duration $transition-function;

  &::before {
    background-color: var(--secondary-color);
    bottom: 0;
    content: '';
    height: 0.4rem;
    left: 0;
    position: absolute;
    right: 0;
    transform: scaleX(0);
    transition: transform $quick-transition-duration $transition-function;
  }

  &:hover::before {
    transform: scaleX(1);
  }

  &:active,
  &[data-active='true'] {
    background-color: var(--secondary-color);
    color: var(--navbar-text-color);
    transform: scale(0.98);
  }
}

@media only screen and (max-width: $break-point) {
  .navbar-item {
    min-width: 25%;
    overflow: hidden;
    padding: 0.8rem 0.2rem;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
}
