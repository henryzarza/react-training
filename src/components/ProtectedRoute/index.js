import React, { lazy, createContext, useContext } from 'react';

const Login = lazy(() => import('../../screens/Login'));

export const AuthContext = createContext();

const ProtectedRoute = ({ component: Component, ...rest }) => {
  const { isAuth } = useContext(AuthContext);

  return isAuth ? <Component {...rest} /> : <Login />;
};

export default ProtectedRoute;
