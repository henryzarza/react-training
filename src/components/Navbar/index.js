import React, { useCallback, useContext } from 'react';
import i18next from 'i18next';
import { Link } from '@reach/router';

import { NAVBAR_ROUTES } from '@constants/routes';
import { AuthContext } from '../ProtectedRoute';
import styles from './styles.module.scss';

function Navbar() {
  const { handleSetIsAuth } = useContext(AuthContext);

  const isLinkActive = useCallback(
    ({ isCurrent }) =>
      // eslint-disable-next-line no-useless-computed-key
      isCurrent ? { ['data-active']: true } : {},
    []
  );

  const handleLogout = useCallback(() => handleSetIsAuth(false), [
    handleSetIsAuth,
  ]);

  return (
    <>
      <section className={styles.content} tabIndex='-1'>
        <h2 className={`title m-bottom-3 ${styles.title}`}>
          {i18next.t('LOGIN:TITLE')}
        </h2>
        <p className={`base-text m-bottom-6 ${styles.paragraph}`}>
          {i18next.t('NAVBAR:DESCRIPTION')}
        </p>
        <a
          rel='nofollow external'
          target='blank'
          href='https://github.com/lennertVanSever/graphcountries'
          className='small-text link m-bottom-2 fw-bold'
        >
          {i18next.t('NAVBAR:REPOSITORY')} lennertVanSever/graphcountries
        </a>
        <small className='small-text m-bottom-1'>
          {i18next.t('NAVBAR:MADE_TEXT')}
          <span className='m-left-1' role='img' aria-label='love'>
            ♥️
          </span>
          {i18next.t('NAVBAR:MADE_BY')}
        </small>
      </section>
      <nav className={styles.navbar}>
        {NAVBAR_ROUTES.map((el) => (
          <Link
            key={el.route}
            to={el.route}
            className={`big-text fw-thin uppercase-text text-center ${styles.navbarItem}`}
            getProps={isLinkActive}
          >
            {i18next.t(el.text)}
          </Link>
        ))}
        <button
          type='button'
          className={`big-text fw-thin uppercase-text text-center ${styles.navbarItem}`}
          onClick={handleLogout}
        >
          {i18next.t('NAVBAR:LOGOUT')}
        </button>
      </nav>
    </>
  );
}

export default Navbar;
