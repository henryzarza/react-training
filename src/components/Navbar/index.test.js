import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { AuthContext } from '@components/ProtectedRoute';
import Navbar from './index';

afterEach(cleanup);

describe('Navbar', () => {
  test('it renders well', () => {
    const { container } = render(
      <AuthContext.Provider value={{ handleSetIsAuth: jest.fn() }}>
        <Navbar />
      </AuthContext.Provider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('it logout correct', async () => {
    const handleSetIsAuth = jest.fn();
  
    render(
      <AuthContext.Provider value={{handleSetIsAuth}}>
        <Navbar />
      </AuthContext.Provider>
    );

    userEvent.click(screen.getByRole('button', { name: 'NAVBAR:LOGOUT' }));

    expect(handleSetIsAuth).toHaveBeenCalledWith(false);
  });
});
