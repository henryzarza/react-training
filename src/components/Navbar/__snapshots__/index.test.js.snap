// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar it renders well 1`] = `
<section
  class="content"
  tabindex="-1"
>
  <h2
    class="title m-bottom-3 title"
  >
    LOGIN:TITLE
  </h2>
  <p
    class="base-text m-bottom-6 paragraph"
  >
    NAVBAR:DESCRIPTION
  </p>
  <a
    class="small-text link m-bottom-2 fw-bold"
    href="https://github.com/lennertVanSever/graphcountries"
    rel="nofollow external"
    target="blank"
  >
    NAVBAR:REPOSITORY
     lennertVanSever/graphcountries
  </a>
  <small
    class="small-text m-bottom-1"
  >
    NAVBAR:MADE_TEXT
    <span
      aria-label="love"
      class="m-left-1"
      role="img"
    >
      ♥️
    </span>
    NAVBAR:MADE_BY
  </small>
</section>
`;
